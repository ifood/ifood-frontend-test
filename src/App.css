.App {
  text-align: center;
}

html, body {
  background-color: white;
}

body {
  padding: 10px;
}

