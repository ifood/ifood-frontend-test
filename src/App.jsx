/* eslint-disable react/prefer-stateless-function */
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';

import { Component } from 'react';
import Home from './pages/Home';
import Login from './pages/Login';

import store from './store';

import GlobalStyles from './styles/globalStyle';

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <GlobalStyles />
        <Router>
          <Switch>
            <Route path="/" exact component={Login} />
            <Route path="/home" component={Home} />
          </Switch>
        </Router>
      </Provider>
    );
  }
}

export default App;
