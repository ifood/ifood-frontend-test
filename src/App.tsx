import React from "react";
import GlobalStyle from "./styles/global";
import Main from "./pages/Main";
import { ToastProvider } from "./hooks/toast";

const App: React.FC = () => (
  <>
    <ToastProvider>
      <Main />
    </ToastProvider>
    <GlobalStyle />
  </>
);

export default App;
