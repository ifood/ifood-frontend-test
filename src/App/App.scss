.app-component {
  display: flex;
  flex-direction: column;
  width: 100%;
  max-height: 100vh;
}
