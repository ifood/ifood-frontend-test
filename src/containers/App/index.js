import React from 'react'
import { Switch, Route } from 'react-router-dom'
import { ThemeProvider } from 'styled-components'

import HomePage from '../HomePage'
import NotFoundPage from '../NotFoundPage'
import PlaylistsPage from '../PlaylistsPage'

import Header from '../../components/Header'
import Footer from '../../components/Footer'

import { HOME_PATH } from './urls'
import './styles.css'
import { theme } from './theme'

function App() {
  return (
    <ThemeProvider theme={theme}>
      <div className="App">
        <Header />

        <div className="App-container">
          <Switch>
            <Route exact path="/" component={HomePage} />
            <Route path={HOME_PATH} component={PlaylistsPage} />
            <Route component={NotFoundPage} />
          </Switch>
        </div>

        <Footer />
      </div>
    </ThemeProvider>
  )
}

export default App
