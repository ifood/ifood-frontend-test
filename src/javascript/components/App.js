import React from 'react';
import Filter from '../containers/filter';
import Login from '../containers/Login';
import { Provider } from "react-redux";
import store from "../store";

class App extends React.Component {

  render() {
    return (
      <Provider store={store}>
        <div className="App">
          <Login />
          <Filter />
        </div>
      </Provider>
    );
  }
}

export default App;
