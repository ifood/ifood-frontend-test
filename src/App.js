import React, { Component } from 'react';
import './App.css';
import PlaylistCounter from './components/playlist-counter/playlist-counter';
import TotalTracks from './components/total-tracks/total-tracks';
import Filter from './components/filter/filter';
import Playlist from './components/playlist/playlist';
import { URL_ME, URL_FEATURED_PLAYLIST, URL_HEROKU, TOKEN_LOCAL_STORAGE,
  URL_AUTHORIZE, INTERVAL, SPOTIFY_API_ERROR_MESSAGE } from './constants';
import logo from './images/SpotifoodIcon.jpg';
import Axios from 'axios';

class App extends Component {
  localeTemp = '';
  countryTemp = '';
  datetimeTemp = '';
  limitTemp = '';
  offsetTemp = '';

  constructor() {
    super();
    this.state = {
      filterString: '',
      intervalId: null,
      filters: {
        locale: '',
        country: '',
        datetime: '',
        limit: '',
        offset: ''
      }
    }
  }

  // Unmount cycle life
  componentWillUnmount() {
    clearInterval(this.state.intervalId);
  }

  // Method to call Spotify API with filter applied
  callSpotifyPlaylistService = (filtersTemp) => {
    if (!filtersTemp) { // This check is for refresh page
      filtersTemp = this.state.filters;
    }

    let accessToken = localStorage.getItem(TOKEN_LOCAL_STORAGE); // Get the token from local storage
    if (accessToken) {
      // Mount the url filter
      const urlFilterFinal = this.buildFilterUrl(filtersTemp);

      // Get the new Playlist using the filter fields
      this.getPlaylistFromSpotify(urlFilterFinal, accessToken);
     }
  }

   // Method to build the filter url for spotify api
  buildFilterUrl(filtersTemp) {
    let urlFilter = URL_FEATURED_PLAYLIST;
    let hasAtLeastOne = false;

    if (filtersTemp.country) {
      let ctry = "";
      if (filtersTemp.country === "en_US") { // This is a bug in mocky API
        ctry = "US";
      } else {
        ctry = filtersTemp.country;
      }
      urlFilter = urlFilter + "?country=" + ctry;
      hasAtLeastOne = true;
    }

    if (filtersTemp.locale) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&locale=" + filtersTemp.locale;
      } else {
        urlFilter = urlFilter + "?locale=" + filtersTemp.locale;
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.datetime) {
      let pattern = ":";
      const properlyTimestamp = filtersTemp.datetime.replace(pattern, "%3A");
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&timestamp=" + properlyTimestamp + "%3A00";
      } else {
        urlFilter = urlFilter + "?timestamp=" + properlyTimestamp + "%3A00";
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.limit && parseInt(filtersTemp.limit, 10) >= 1 && parseInt(filtersTemp.limit, 10) <= 50) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&limit=" + filtersTemp.limit;
      } else {
        urlFilter = urlFilter + "?limit=" + filtersTemp.limit;
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.offset && parseInt(filtersTemp.offset, 10) >= 1) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&offset=" + filtersTemp.offset;
      } else {
        urlFilter = urlFilter + "?offset=" + filtersTemp.offset;
      }
    }
    return urlFilter;
  }

  // Method to refresh page after 30 seconds
  refreshPage = () => {
    const intervalId = setInterval(this.callSpotifyPlaylistService, INTERVAL);
    this.setState({
      intervalId,
    });
  }

  // Method called when the user change some information on filter
  filterList = (event) => {
    const field = {
      [event.target.name]: event.target.value,
    };

    if (field.locale) {
      this.localeTemp = field.locale;
    } else if (field.country) {
      this.countryTemp = field.country;
    } else if (field.datetime || field.datetime === "") {
      this.datetimeTemp = field.datetime;
    } else if (field.limit) {
      if (parseInt(field.limit, 10) < 1 || parseInt(field.limit, 10) > 50) {
        this.limitTemp = "";
      } else {
        this.limitTemp = field.limit;
      }
    } else {
      if (parseInt(field.offset, 10) < 1) {
        this.offsetTemp = "";
      } else {
        this.offsetTemp = field.offset;
      }
    }

    this.setState({
      filters: {
        locale: this.localeTemp ? this.localeTemp : '',
        country: this.countryTemp ? this.countryTemp : '',
        datetime: this.datetimeTemp ? this.datetimeTemp : '',
        limit: this.limitTemp ? this.limitTemp : '',
        offset: this.offsetTemp ? this.offsetTemp : ''
      }
    });

    const filtersTemp = {
      locale: this.localeTemp ? this.localeTemp : '',
      country: this.countryTemp ? this.countryTemp : '',
      datetime: this.datetimeTemp ? this.datetimeTemp : '',
      limit: this.limitTemp ? this.limitTemp : '',
      offset: this.offsetTemp ? this.offsetTemp : ''
    };

    this.callSpotifyPlaylistService(filtersTemp);
  }

  // Method to call Spotify Login in case the user is not logged already
  openSpotifyLogin() {
      let url = URL_AUTHORIZE;
          url += '?client_id=' + encodeURIComponent(process.env.REACT_APP_SPOTIFY_CLIENT_ID);
          url += '&response_type=token';
          url += '&redirect_uri=' + encodeURIComponent(URL_HEROKU);
      window.location = url;
  }

  getUserNameInformationFromSpotify(accessToken) {
    Axios.get(URL_ME, {
      headers: { 'Authorization': 'Bearer ' + accessToken }
    }).then(response =>
        this.setState({
        user: {
          name: response.data.display_name
          }
        })
    )
    .catch(error => {
        console.log(SPOTIFY_API_ERROR_MESSAGE, error)
    });
  }

  getPlaylistFromSpotify(url, accessToken) {
    Axios.get(url, {
      headers: { 'Authorization': 'Bearer ' + accessToken }
    }).then(playlistData => {
        if (!playlistData.error) {
        let playlists = playlistData.data.playlists.items;
          this.setState({
            playlists: playlists.map(item => {
              return {
                name: item.name,
                imageUrl: item.images[0].url,
                numberOfTracks: item.tracks.total,
                owner: item.owner.display_name
              }
            })
          })
        } else {
          this.setState({
            playlists: []
          })
        }
      })
      .catch(error => {
        this.setState({
            playlists: []
          })
           console.log(SPOTIFY_API_ERROR_MESSAGE, error)
      });
  }

  // Method called before Render the page
  componentDidMount() {
    const url = window.location.hash;
    if (url) { // Check if the access token is on url
      const access_token = url.match(/\#(?:access_token)\=([\S\s]*?)\&/)[1]; // Get only the access token from url
      localStorage.setItem(TOKEN_LOCAL_STORAGE, access_token); // Save the token on local storage
      window.location.href = '/'; // Clean the URL to not be exposed to user (access token)
    }

    let accessToken = localStorage.getItem(TOKEN_LOCAL_STORAGE); // Check if already have a token into local storage
    if (!accessToken) {
      return;
    }

    // Call this API to get user name information
    this.getUserNameInformationFromSpotify(accessToken);

    // Get all Feature Playlists without filter
    this.getPlaylistFromSpotify(URL_FEATURED_PLAYLIST, accessToken);

    // Start the an interval of 30 seconds to refresh the playlist
    this.refreshPage();
  }

  // Method to render the page
  render() {
    // If there is no user and playlist, return empty array
    let playlistToShow = this.state.user && this.state.playlists ?
      this.state.playlists.filter(playlist => {
        // Here I apply the filter for Playlist name
        let matchesPlaylist = playlist.name.toLowerCase()
        .includes(this.state.filterString.toLocaleLowerCase());
        return matchesPlaylist;
      }) : []

    return (
      <div className="App">
        {
          this.state.user ?
            <div>
              <img src={logo} alt="Spotifood"></img>
              <h2>User logged in: <span style={{color: 'green', fontWeight: 'bold'}}>{this.state.user.name}</span></h2>
              <PlaylistCounter playlists={playlistToShow} />
              <TotalTracks playlists={playlistToShow} />
              <Filter onTextChange={text => this.setState({ filterString: text })}
                onChangeFilters={this.filterList}
                filterValues={this.state.filters} />
              {
                playlistToShow.length > 0 ?
                  playlistToShow.map((playlist, i) => <Playlist key={i} playlist={playlist} index={i} />)
                  : <div style={{ color: 'red', fontWeight: 'bold', marginTop: '50px' }}>No Playlist</div>
              }
            </div> : <button onClick={() => {
                this.openSpotifyLogin();
            }
            }
              style={{ padding: '', fontSize: '50px', marginTop: '20px' }}>Sign in with Spotify</button>
        }
      </div>
    );
  }

}

export default App;