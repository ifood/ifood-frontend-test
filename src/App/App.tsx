import React from 'react';
import { BrowserRouter, Route, Redirect, RouteComponentProps } from 'react-router-dom';

import './App.scss';

import LandingPage from '../pages/LandingPage/LandingPage';
import { SpotifyServiceAuth } from '../services/auth.service';
import Playlists from '../pages/Playlists/Playlists';

const PrivateRoute = ({ Component, path }: { Component: React.FC<RouteComponentProps>, path: string }) => (
  <Route path={path} render={props => (
    SpotifyServiceAuth.hasToken() ? <Component {...props} /> : <Redirect to="/" />
  )} />
);

const App = () => {
  return (
    <div className="app-component">
      <BrowserRouter>
        <Route path="/" exact component={LandingPage} />
        <PrivateRoute path="/playlists" Component={Playlists} />
      </BrowserRouter>
    </div>
  );
}

export default App;
