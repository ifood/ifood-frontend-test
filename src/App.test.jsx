import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from './App';

jest.mock('./core/AuthorizationContext', () => ({
  __esModule: true,
  default: {
    /* eslint-disable-next-line  */
    Provider: ({ children }) => <div>{children}</div>,
  },
}));

describe('<App />', () => {
  it('snapshot', () => {
    const wrapper = mount(<App />);
    jestExpect(toJson(wrapper)).toMatchSnapshot();
  });
});
