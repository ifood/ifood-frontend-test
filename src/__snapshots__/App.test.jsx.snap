// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot 1`] = `
<App>
  <Provider
    value={null}
  >
    <div>
      <IntlProvider
        locale="en-US"
        messages={
          Object {
            "cancel": "Cancel",
            "choose.filters.message": "Choose new filters to find amazing playlists",
            "datetime.placeholder": "ex: 12/22/2020, 12:00 AM",
            "disclaimer": "Spotifood is not a real brand and has no commercial purposes, all of its content belongs to Spotify. The creation and use of the brand are limited to this app, which has the sole purpose of being a technical test",
            "disclaimer.data": "We don't store any Spotify data",
            "featured.playlists": "Featured playlists",
            "filter": "Filter",
            "filter.submit": "Submit",
            "filters": "Filters",
            "loading": "Loading",
            "login.message": "Hello, log in with your Spotify account to access amazing Playlists.",
            "login.on.spotify": "Login on Spotify",
            "no.playlist": "No Playlist found",
            "ops": "Ops!",
            "playlist.tracks": "{n} {n, plural, one {track} other {tracks}}",
            "playlists": "Playlists",
            "search": "Search",
            "time": "Time",
          }
        }
      >
        <div>
          <BrowserRouter>
            <Router
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
            >
              <Routes>
                <Switch>
                  <AuthRoute
                    computedMatch={
                      Object {
                        "isExact": true,
                        "params": Object {},
                        "path": "/",
                        "url": "/",
                      }
                    }
                    fallbackRoute={<Login />}
                    location={
                      Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      }
                    }
                    path="/"
                  >
                    <Route
                      computedMatch={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/",
                          "url": "/",
                        }
                      }
                      location={
                        Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        }
                      }
                      path="/"
                      render={[Function]}
                    >
                      <Login>
                        <styled.div>
                          <div
                            className="sc-fzqBZW ipbYhy"
                          >
                            <Header>
                              <styled.header>
                                <header
                                  className="sc-fznKkj lfPylV"
                                >
                                  <styled.h1>
                                    <h1
                                      className="sc-fznZeY kbIVGi"
                                    >
                                      <styled.img
                                        alt="Spotifood logo"
                                        src="logo.png"
                                      >
                                        <img
                                          alt="Spotifood logo"
                                          className="sc-fzokOt llqEFR"
                                          src="logo.png"
                                        />
                                      </styled.img>
                                    </h1>
                                  </styled.h1>
                                </header>
                              </styled.header>
                            </Header>
                            <styled.main>
                              <main
                                className="sc-AxmLO bdGGKO"
                              >
                                <styled.div>
                                  <div
                                    className="sc-fzoLag cVoaIU"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-fzoXzr gFmAEj"
                                      >
                                        <styled.img
                                          alt="spotifood banner"
                                          src="banner.png"
                                        >
                                          <img
                                            alt="spotifood banner"
                                            className="sc-fzoyAV jmIzDp"
                                            src="banner.png"
                                          />
                                        </styled.img>
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzpjYC jvcZLQ"
                                      >
                                        <styled.h2>
                                          <h2
                                            className="sc-fzqNJr bYnXfq"
                                          >
                                            <FormattedMessage
                                              id="login.message"
                                            />
                                          </h2>
                                        </styled.h2>
                                        <Button
                                          as="button"
                                          color="red"
                                          onClick={[Function]}
                                        >
                                          <Ref
                                            innerRef={
                                              Object {
                                                "current": <button
                                                  class="ui red button"
                                                />,
                                              }
                                            }
                                          >
                                            <RefFindNode
                                              innerRef={
                                                Object {
                                                  "current": <button
                                                    class="ui red button"
                                                  />,
                                                }
                                              }
                                            >
                                              <button
                                                className="ui red button"
                                                onClick={[Function]}
                                              >
                                                <FormattedMessage
                                                  id="login.on.spotify"
                                                />
                                              </button>
                                            </RefFindNode>
                                          </Ref>
                                        </Button>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </main>
                            </styled.main>
                            <Footer>
                              <styled.footer>
                                <footer
                                  className="sc-fzozJi ixBtCB"
                                >
                                  <styled.div>
                                    <div
                                      className="sc-fzoLsD dSybVf"
                                    >
                                      <styled.div>
                                        <div
                                          className="sc-fzpans bzPxPR"
                                        >
                                          <styled.img
                                            alt="Spotifood logo"
                                            src="logo-icon.png"
                                          >
                                            <img
                                              alt="Spotifood logo"
                                              className="sc-fzplWN hdFbkJ"
                                              src="logo-icon.png"
                                            />
                                          </styled.img>
                                        </div>
                                      </styled.div>
                                      <styled.div>
                                        <div
                                          className="sc-fznyAO idjYWI"
                                        >
                                          <p>
                                            <FormattedMessage
                                              id="disclaimer"
                                            />
                                          </p>
                                          <p>
                                            <FormattedMessage
                                              id="disclaimer.data"
                                            />
                                          </p>
                                        </div>
                                      </styled.div>
                                    </div>
                                  </styled.div>
                                </footer>
                              </styled.footer>
                            </Footer>
                          </div>
                        </styled.div>
                      </Login>
                    </Route>
                  </AuthRoute>
                </Switch>
              </Routes>
            </Router>
          </BrowserRouter>
        </div>
      </IntlProvider>
    </div>
  </Provider>
</App>
`;
